spec/pr_size_checker_spec.rb

```diff
@@ -0,0 +1,194 @@
+# frozen_string_literal: true
+
+require_relative 'spec_helper'
+
+module Danger
+  describe Danger::PRSizeChecker do
+    it 'is a plugin' do
+      expect(described_class.new(nil)).to be_a Danger::Plugin
+    end
+
+    describe 'with Dangerfile' do
+      before do
+        @dangerfile = testing_dangerfile
+        @plugin = @dangerfile.pr_size_checker
+      end
+
+      context 'when checking a PR diff size' do
+        before do
+          allow(@plugin.git).to receive(:added_files).and_return([])
+          allow(@plugin.git).to receive(:modified_files).and_return([])
+          allow(@plugin.git).to receive(:deleted_files).and_return([])
+        end
+
+        shared_examples 'using the default diff size counter without a file selector' do |type|
+          let(:diff_counter_for_type) do
+            type_hash = {
+              insertions: :insertions,
+              deletions: :deletions,
+              all: :lines_of_code
+            }
+
+            type_hash[type]
+          end
+
+          it 'returns a warning when using default parameters in a PR that has larger diff than the default maximum' do
+            allow(@plugin.git).to receive(diff_counter_for_type).and_return(501)
+
+            @plugin.check_diff_size(type: type)
+
+            expect(@dangerfile.status_report[:warnings]).to eq ['Please keep the Pull Request small, breaking it down into multiple ones if necessary']
+          end
+
+          it 'does nothing when using default parameters in a PR that has equal diff than the default maximum' do
+            allow(@plugin.git).to receive(diff_counter_for_type).and_return(500)
+
+            @plugin.check_diff_size(type: type)
+
+            expect(@dangerfile.status_report[:errors]).to be_empty
+            expect(@dangerfile.status_report[:warnings]).to be_empty
+          end
+
+          it 'does nothing when using default parameters in a PR that has smaller diff than the default maximum' do
+            allow(@plugin.git).to receive(diff_counter_for_type).and_return(499)
+
+            @plugin.check_diff_size(type: type)
+
+            expect(@dangerfile.status_report[:errors]).to be_empty
+            expect(@dangerfile.status_report[:warnings]).to be_empty
+          end
+
+          it 'returns an error using a custom message and a custom pr body size' do
+            allow(@plugin.git).to receive(diff_counter_for_type).and_return(600)
+
+            custom_message = 'diff size custom message'
+            @plugin.check_diff_size(type: type, max_size: 599, message: custom_message, fail_on_error: true)
+
+            expect(@dangerfile.status_report[:errors]).to eq [custom_message]
+          end
```

Comment:
This test looks like it's asserting two distinct behaviors: that a custom message is used when given, and that when we `fail_on_error: true` an error is reported.

Would you consider splitting into multiple `it`s? I'm thinking

- it reports a warning with a custom message when one is given
- it reports a warning when fail on error is false
- it reports an error when fail on error is true
- it reports an error with a custom message when one is given and fail on error is true (optional)

Reply:
I had the same thought when I first implemented something like this, but I wasn't sure about the added value of actually having separate checks for each of the combinations (given they're relatively simple should be covered in other tests).
But I do agree that in general it is a good practice to have single-purpose, isolated tests.

I gave it a try with more shared examples to avoid the repetition, please let me know what you think: 18b567b2520a2244ce3aa52bbe5016656eff5cf2